polkadot/runtime/parachains/src/scheduler.rs

```diff
@@ -531,183 +321,150 @@ impl<T: Config> Pallet<T> {
-			.scheduler_params
-			.max_availability_timeouts;
-		Self::next_up_on_available(core).or_else(|| {
-			// Or, if none, the claim currently occupying the core,
-			// as it would be put back on the queue after timing out if number of retries is not at
-			// the maximum.
-			let cores = AvailabilityCores::<T>::get();
-			cores.get(core.0 as usize).and_then(|c| match c {
-				CoreOccupied::Free => None,
-				CoreOccupied::Paras(pe) =>
-					if pe.availability_timeouts < max_availability_timeouts {
-						Some(Self::paras_entry_to_scheduled_core(pe))
-					} else {
-						None
-					},
-			})
-		})
-	}
+	/// For each core that isn't part of the `except_for` set, pop the first item of the claim queue
+	/// and fill the queue from the assignment provider.
+	pub(crate) fn advance_claim_queue(except_for: &BTreeSet<CoreIndex>) {
+		let config = configuration::ActiveConfig::<T>::get();
+		let num_assigner_cores = config.scheduler_params.num_cores;
+		// Extra sanity, config should already never be smaller than 1:
+		let n_lookahead = config.scheduler_params.lookahead.max(1);
+
+		for core_idx in 0..num_assigner_cores {
+			let core_idx = CoreIndex::from(core_idx);
+
+			if !except_for.contains(&core_idx) {
+				let core_idx = CoreIndex::from(core_idx);
 
-	/// Pushes occupied cores to the assignment provider.
-	fn push_occupied_cores_to_assignment_provider() {
-		AvailabilityCores::<T>::mutate(|cores| {
-			for core in cores.iter_mut() {
-				match core::mem::replace(core, CoreOccupied::Free) {
-					CoreOccupied::Free => continue,
-					CoreOccupied::Paras(entry) => {
-						Self::maybe_push_assignment(entry);
-					},
+				if let Some(dropped_para) = Self::pop_front_of_claim_queue(&core_idx) {
+					T::AssignmentProvider::report_processed(dropped_para);
 				}
-			}
-		});
-	}
 
-	// on new session
-	fn push_claim_queue_items_to_assignment_provider() {
-		for (_, claim_queue) in ClaimQueue::<T>::take() {
-			// Push back in reverse order so that when we pop from the provider again,
-			// the entries in the claim queue are in the same order as they are right now.
-			for para_entry in claim_queue.into_iter().rev() {
-				Self::maybe_push_assignment(para_entry);
+				Self::fill_claim_queue(core_idx, n_lookahead);
 			}
 		}
 	}
 
-	/// Push assignments back to the provider on session change unless the paras
-	/// timed out on availability before.
-	fn maybe_push_assignment(pe: ParasEntryType<T>) {
-		if pe.availability_timeouts == 0 {
-			T::AssignmentProvider::push_back_assignment(pe.assignment);
+	/// Get an iterator into the claim queues.
+	///
+	/// This iterator will have an item for each and every core index up to the maximum core index
+	/// found in the claim queue. In other words there will be no holes/missing core indices,
+	/// between core 0 and the maximum, even if the claim queue was missing entries for particular
+	/// indices in between. (The iterator will return an empty `VecDeque` for those indices.
```

Comment:
agree. Now that you bring this up, why do we want to fill in missing entries with empty vecdeques?
With the current code I don't see the difference it brings. Is there any?

Reply:
I think it was needed before because we were `zip`-ing with the availability cores in `eligible_paras`. Since we no longer do that, I think I can remove it

Reply:
removed it